src/cryptonote_core/service_node_list.cpp

```diff
@@ -442,14 +482,46 @@ namespace service_nodes
     return expired_nodes;
   }
 
+  uint64_t service_node_list::get_min_contribution(uint64_t staking_requirement) const
+  {
+    return staking_requirement / 10;
+  }
+
   std::vector<std::pair<cryptonote::account_public_address, uint32_t>> service_node_list::get_winner_addresses_and_shares(const crypto::hash& prev_id) const
   {
     crypto::public_key key = select_winner(prev_id);
     if (key == crypto::null_pkey)
       return { std::make_pair(null_address, STAKING_SHARES) };
     std::vector<std::pair<cryptonote::account_public_address, uint32_t>> winners;
+    uint64_t remaining_shares = STAKING_SHARES;
+    auto add_to_winners = [&winners](cryptonote::account_public_address address, uint64_t amount_of_shares) {
+      if (amount_of_shares == 0) return;
+      auto iter = std::find_if(std::begin(winners), std::end(winners),
+          [&address](const std::pair<cryptonote::account_public_address, uint32_t>& winner) {
+            return winner.first == address;
+          });
+      if (iter == winners.end())
+        winners.push_back(std::make_pair(address, amount_of_shares));
+      else
+        iter->second += amount_of_shares;
+    };
     for (size_t i = 0; i < m_service_nodes_infos.at(key).addresses.size(); i++)
-      winners.push_back(std::make_pair(m_service_nodes_infos.at(key).addresses[i], m_service_nodes_infos.at(key).shares[i]));
+    {
+      add_to_winners(m_service_nodes_infos.at(key).addresses[i], m_service_nodes_infos.at(key).shares[i]);
+      remaining_shares -= m_service_nodes_infos.at(key).shares[i];
+    }
+    uint64_t remaining_staking_requirement = m_service_nodes_infos.at(key).staking_requirement;
+    for (const auto& contribution : m_service_nodes_infos.at(key).contributions)
+    {
+      uint64_t actual_contribution = std::min(remaining_staking_requirement, contribution.amount);
+      remaining_staking_requirement -= actual_contribution;
+
```

Comment:
assert here? I don't think it's possible for this to underflow since process_contribution_tx() has the is_fully_funded() check to stop excess contributions from coming through. But having it here would be good for documentation and indicate that someone has checked this and has reasonable expectation that it won't break.

Reply:
The logic for this not underflowing is local to the code here. actual_contribution is at most remaining_staking_requirement, see line 516.

A comment here would not help much

Reply:
lol. Took me one nights worth of sleep to realise that. Sorry, this was moot.

Reply:
It's ok